Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp

```diff
@@ -0,0 +1,247 @@
+				UE_LOG(LogBenchmarkGymGameModeBase, Error, TEXT("A client connection was dropped. Expected %d, got %d"), ExpectedPlayers, GetNumPlayers());
+			}
+			else
+			{
+				// Useful for NFR log inspection
+				UE_LOG(LogBenchmarkGymGameModeBase, Log, TEXT("All clients successfully connected. Expected %d, got %d"), ExpectedPlayers, GetNumPlayers());
+			}
+		}
+	}
+}
+
+void ABenchmarkGymGameModeBase::TickFPSCheck(float DeltaSeconds)
+{
+	if (MinDelayFPS > 0.0f)
+	{
+		MinDelayFPS -= DeltaSeconds;
+	}
+
+	if (MinDelayFPS <= 0.0f && !bHasFpsFailed && GetWorld() != nullptr)
+	{
+		if (const UGDKTestGymsGameInstance* GameInstance = Cast<UGDKTestGymsGameInstance>(GetWorld()->GetGameInstance()))
+		{
+			float FPS = GameInstance->GetAveragedFPS();
+			if (FPS < MinAcceptableFPS)
+			{
+				bHasFpsFailed = true;
+#if !WITH_EDITOR
+				UE_LOG(LogBenchmarkGymGameModeBase, Log, TEXT("FPS check failed. FPS: %.8f"), FPS);
+#endif
+			}
+		}
+	}
+}
+
+void ABenchmarkGymGameModeBase::TickUXMetricCheck(float DeltaSeconds)
+{
+	if (!HasAuthority())
+	{
+		return;
+	}
+
+	float ClientRTTSeconds = 0.0f;
+	int UXComponentCount = 0;
+	float ClientViewLatenessSeconds = 0.0f;
+	for (TObjectIterator<UUserExperienceReporter> Itr; Itr; ++Itr) // These exist on player characters
+	{
+		UUserExperienceReporter* Component = *Itr;
+		if (Component->GetOwner() != nullptr && Component->GetWorld() == GetWorld())
+		{
+			ClientRTTSeconds += Component->ServerRTT;
+			UXComponentCount++;
+
+			ClientViewLatenessSeconds += Component->ServerViewLateness;
+		}
+	}
+
+	ActivePlayers = UXComponentCount;
+
+	if (UXComponentCount > 0)
+	{
+		bPlayersHaveJoined = true;
+	}
+
+	if (!bPlayersHaveJoined)
+	{
+		return; // We don't start reporting until there are some valid components in the scene.
+	}
+
+	ClientRTTSeconds /= static_cast<float>(UXComponentCount) + 0.00001f; // Avoid div 0
+	ClientViewLatenessSeconds /= static_cast<float>(UXComponentCount) + 0.00001f; // Avoid div 0
+
+	AveragedClientRTTSeconds = ClientRTTSeconds;
+	AveragedClientViewLatenessSeconds = ClientViewLatenessSeconds;
+
+	if (!bHasUxFailed && AveragedClientRTTSeconds > MaxClientRoundTripSeconds && AveragedClientViewLatenessSeconds > MaxClientViewLatenessSeconds)
+	{
+		bHasUxFailed = true;
+#if !WITH_EDITOR
+		UE_LOG(LogBenchmarkGymGameModeBase, Error, TEXT("UX metric has failed. RTT: %.8f, ViewLateness: %.8f, ActivePlayers: %d"), AveragedClientRTTSeconds, AveragedClientViewLatenessSeconds, ActivePlayers);
+#endif
+	}
+
+	PrintUXMetric -= DeltaSeconds;
+	if (PrintUXMetric < 0.0f)
+	{
+		PrintUXMetric = 10.0f;
+#if !WITH_EDITOR
+		UE_LOG(LogBenchmarkGymGameModeBase, Log, TEXT("UX metric values. RTT: %.8f, ViewLateness: %.8f, ActivePlayers: %d"), AveragedClientRTTSeconds, AveragedClientViewLatenessSeconds, ActivePlayers);
+#endif
+	}
+}
+
+void ABenchmarkGymGameModeBase::ParsePassedValues()
+{
+	if (FParse::Param(FCommandLine::Get(), TEXT("TotalPlayers")))
```

Comment:
Can we reuse the same "OverrideSpawning" arg used elsewhere?

Reply:
We could but I don't think OverrideSpawning is relevant for offloading.

Reply:
I guess I don't like the inference that `TotalPlayers` means you should read the rest of the values from command line. Can we implement a process that can be shared between setups then?

Reply:
Sure so we change OverrideSpawning to something more generic like "ParseCommandLine"?

Reply:
In fact we should have both.